sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -852,26 +852,142 @@ class Analyzer(
   }
 
   /**
-   * This rule resolve subqueries inside expressions.
+   * This rule resolves subqueries inside expressions and it rewrites correlated & scalar
+   * subqueries.
    *
-   * Note: CTE are handled in CTESubstitution.
+   * It works as follows:
+   * 1. Find all subquery expressions in a logical plan. Try to resolve them and update the
+   *    SubQueryExpression with the resolved logical plan.
+   * 2. For Filter (i.e. WHERE/HAVING clauses) the condition will be split by AND. Each part of the
+   *    condition that contains a [[CorrelatedSubqueryExpression]] will be rewritten:
+   *    a. EXISTS/NOT EXISTS will be rewritten as semi/anti join, unresolved conditions in Filter
+   *       will be pulled out as the join conditions.
+   *    b. IN/NOT IN will be rewritten as semi/anti join, unresolved conditions in the Filter will
+   *       be pulled out as join conditions, value = selected column will also be used as join
+   *       condition.
+   *
+   * Note: CTEs are handled in CTESubstitution.
    */
   object ResolveSubquery extends Rule[LogicalPlan] with PredicateHelper {
 
-    private def hasSubquery(e: Expression): Boolean = {
-      e.find(_.isInstanceOf[SubqueryExpression]).isDefined
+    private def hasCorrelatedSubquery(e: Expression): Boolean = {
+      e.find(_.isInstanceOf[CorrelatedSubqueryExpression]).isDefined
     }
 
     private def hasSubquery(q: LogicalPlan): Boolean = {
-      q.expressions.exists(hasSubquery)
+      q.expressions.exists(_.find(_.isInstanceOf[SubqueryExpression]).isDefined)
+    }
+
+    /**
+     * Removes the conjunctive predicates of Filter that can't be resolved in this logical plan,
+     * returns the resolved new logical plan and removed predicates.
+     */
+    private def removeUnresolvedPredicates(q: LogicalPlan): (LogicalPlan, Option[Expression]) = {
```

Comment:
I am still working on this rule. It is quite hard to follow, and I feel this can be improved.